pkg/controller/atlasprivateendpoint/privateendpoint.go

```diff
@@ -0,0 +1,236 @@
+				api.PrivateEndpointReady,
+				workflow.PrivateEndpointFailedToConfigure,
+				errors.New(compositeInterfaceMap.Atlas.ErrorMessage()),
+			)
+		case inProgress:
+			pendingResources = true
+			continue
+		case inAKO && !inAtlas:
+			gcpProjectID := getGCPProjectID(akoPrivateEndpoint, compositeInterfaceMap.AKO.InterfaceID())
+			_, err := r.privateEndpointService.CreatePrivateEndpointInterface(
+				ctx.Context,
+				projectID,
+				akoPrivateEndpoint.Spec.Provider,
+				akoPrivateEndpoint.Status.ServiceID,
+				gcpProjectID,
+				compositeInterfaceMap.AKO,
+			)
+			if err != nil {
+				return r.terminate(ctx, akoPrivateEndpoint, atlasPEService, api.PrivateEndpointReady, workflow.PrivateEndpointFailedToCreate, err)
+			}
+			pendingResources = true
+		case !inAKO && inAtlas:
+			err := r.privateEndpointService.DeleteEndpointInterface(
+				ctx.Context,
+				projectID,
+				akoPrivateEndpoint.Spec.Provider,
+				akoPrivateEndpoint.Status.ServiceID,
+				compositeInterfaceMap.Atlas.InterfaceID(),
+			)
+			if err != nil {
+				return r.terminate(ctx, akoPrivateEndpoint, atlasPEService, api.PrivateEndpointReady, workflow.PrivateEndpointFailedToDelete, err)
+			}
+			pendingResources = true
+		}
+	}
+
+	if pendingResources {
+		return r.inProgress(ctx, akoPrivateEndpoint, atlasPEService, api.PrivateEndpointReady, workflow.PrivateEndpointUpdating, "Private Endpoints are being updated")
+	}
+
+	return r.ready(ctx, akoPrivateEndpoint, atlasPEService)
+}
+
+// getOrMatchPrivateEndpointService retrieve the project by ID if one is set or try to match by provider/region
+// only one private endpoint service per provider/region is allowed
+func (r *AtlasPrivateEndpointReconciler) getOrMatchPrivateEndpointService(
+	ctx context.Context,
+	projectID string,
+	akoPrivateEndpoint *akov2.AtlasPrivateEndpoint,
+) (privateendpoint.EndpointService, error) {
+	if akoPrivateEndpoint.Status.ServiceID != "" {
+		return r.privateEndpointService.GetPrivateEndpoint(ctx, projectID, akoPrivateEndpoint.Spec.Provider, akoPrivateEndpoint.Status.ServiceID)
+	}
+
+	endpointServices, err := r.privateEndpointService.ListPrivateEndpoints(ctx, projectID, akoPrivateEndpoint.Spec.Provider)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, endpointService := range endpointServices {
+		if endpointService.Region() == akoPrivateEndpoint.Spec.Region {
+			return endpointService, err
+		}
+	}
+
+	return nil, nil
+}
+
+func (r *AtlasPrivateEndpointReconciler) deletePrivateEndpoint(
+	ctx context.Context,
+	projectID string,
+	atlasPEService privateendpoint.EndpointService,
+) (privateendpoint.EndpointService, error) {
+	if len(atlasPEService.EndpointInterfaces()) == 0 && atlasPEService.Status() != privateendpoint.StatusDeleting {
+		err := r.privateEndpointService.DeleteEndpointService(ctx, projectID, atlasPEService.Provider(), atlasPEService.ServiceID())
+		if err != nil {
+			return atlasPEService, err
+		}
+	}
+
+	for _, i := range atlasPEService.EndpointInterfaces() {
```

Comment:
Why is this not before the service deletion? don't we need the endpoints to be deleted first?

Reply:
As there is two pieces moving here and different behaviors depending on the configuration scenario and provider.
This was the best way to handle all case while keep implementation complexity low.
For example, here we ensure we delete a service only when there are not more interfaces pending, moving as you suggest it would require extra conditions in both part to interrupt (early return) the flow.